### ISSUE TITLE
Investigate switch to tch-rs

### ISSUE THREAD

**Issue by danieldk** [stickeritis/sticker#25]
There have been PyTorch bindings in the past, but the `tch-rs` looks to be promising and very active:

https://github.com/LaurentMazare/tch-rs

The main benefit of switching to `tch-rs` would be that we can define networks directly in Rust, making the extra Python detour unnecessary. I like to look first whether all the Tensorflow bits that we are using now have equivalents in `tch-rs`. If `tch-rs` is complete enough, I'll do an experimental implementation.

**Comment by danieldk** [comment]
# General

* Masking: does not seem to be available. We currently hand-roll this in Tensorflow as well.
* Batch normalization: https://docs.rs/tch/0.0.9/tch/struct.Tensor.html#method.batch_norm
* top_k prediction: https://docs.rs/tch/0.0.9/tch/struct.Tensor.html#method.topk

# Bidi RNNs

* LSTM: https://docs.rs/tch/0.0.9/tch/nn/fn.lstm.html
* GRU: https://docs.rs/tch/0.0.9/tch/nn/fn.gru.html
* Bidirectional: https://docs.rs/tch/0.0.9/tch/nn/struct.RNNConfig.html#structfield.bidirectional
* Stacked: https://docs.rs/tch/0.0.9/tch/nn/struct.RNNConfig.html#structfield.num_layers
* Dropout: https://docs.rs/tch/0.0.9/tch/nn/struct.RNNConfig.html#structfield.dropout
* CRF: missing, but does not seem to give an improvment anyway?
* Adam optimizer: https://docs.rs/tch/0.0.9/tch/nn/fn.adam.html

# Dilated convolution

* 1D Convolution: https://docs.rs/tch/0.0.9/tch/nn/fn.conv1d.html
* Dilation: https://docs.rs/tch/0.0.9/tch/nn/struct.ConvConfigND.html#structfield.dilation
* Spatial dropout: Note sure?!? https://docs.rs/tch/0.0.9/tch/struct.Tensor.html#method.feature_dropout -> needs more investigation.
* Weight norm: missing, but we have a custom implementation in Tensorflow as well.

**Comment by danieldk** [comment]
Closing this for now until tch has matured a bit.
